server/src/main/scala/org/ada/server/dataaccess/dataset/SimpleInitializableCache.scala

```diff
@@ -0,0 +1,90 @@
+package org.ada.server.dataaccess.dataset
+
+import collection.mutable.{Map => MMap}
+import scala.concurrent.{Await, Future}
+import play.api.libs.concurrent.Execution.Implicits.defaultContext
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.concurrent.duration._
+
+abstract class SimpleInitializableCache[ID, T](eagerInit: Boolean) {
+
+  private val locks = new ConcurrentHashMap[ID, AnyRef]()
+
+  protected val cache = {
+    val map = MMap[ID, T]()
+    if (eagerInit)
+      initialize(map)
+    map
+  }
+
+  def apply(id: ID): Option[T] =
+    getItemOrElse(id) {
+
+      // get a lock... if doesn't exist, register one
+      val lock =  {
+        val lockAux = locks.putIfAbsent(id, new AnyRef())
+        // putIfAbsent returns 'null' if there was no associated object (in our case a lock) for a given key
+        if (lockAux == null) locks.get(id) else lockAux
+      }
+
+      lock.synchronized {
+        getItemOrElse(id)(
+          createAndCacheInstance(id)
+        )
+      }
+    }
+
+  private def getItemOrElse(
+    id: ID)(
+    initialize: => Option[T]
+  ): Option[T] =
+    cache.get(id) match {
+      case Some(item) => Some(item)
+      case None => initialize
+    }
+
+  private def createAndCacheInstance(id: ID) = {
+    val future = createInstance(id).map { instance =>
+      if (instance.isDefined)
+        cache.put(id, instance.get)
+      instance
+    }
+    // TODO: change to Future
+    Await.result(future, 10 minutes)
```

Comment:
Is there a reason for not returning a Future right now? I'm a bit wary of awaits, especially ones that can last up to 10 minutes. And then there's of course the increased potential for deadlocks.

Reply:
this is the (old) code not affected by this PR... needs to be addressed but simple Future cannot be returned since it's part of initialization. Know how it can be handled though.

Since it's an independent issue will tackle it in ada-discovery/ada-issues#171